modules/service/src/main/scala/lucuma/odb/service/ProgramUserService.scala

```diff
@@ -234,130 +260,170 @@ object ProgramUserService:
+            ) |+| ac
+
+    def deleteProgramUser(
+      pui:             ProgramUser.Id,
+      targetRole:      ProgramUserRole,
+      targetProgramId: Program.Id,
+      sourceUser:      User            // user doing the program user update
+    ): Result[AppliedFragment] =
+      accessCheck("delete", targetRole, targetProgramId, sourceUser)
+        .map:
+          case None     => AppliedFragment.empty
+          case Some(ac) => void" AND " |+| ac
+        .map: ac =>
+          sql"""
+            DELETE FROM t_program_user
+          WHERE c_program_user_id = $program_user_id
+          """(pui) |+| ac
 
-    val UnlinkUnconditionally: Command[(Program.Id, User.Id)] =
+    val SelectLinkData: Query[ProgramUser.Id, (Program.Id, ProgramUserRole, Option[User.Id])] =
       sql"""
-        DELETE FROM
-          t_program_user
+        SELECT
+          c_program_id,
+          c_role,
+          c_user_id
+        FROM t_program_user
         WHERE
-          c_program_id = $program_id
-        AND
-          c_user_id = $user_id
-      """.command
-
-    val UnlinkCoiAsPi: Command[(User.Id, Program.Id, User.Id)] =
-      sql"""
-        DELETE FROM t_program_user
-        WHERE c_program_id = $program_id
-        AND   c_user_id    = $user_id
-        AND   c_role       = 'coi'
-        AND EXISTS (
-          SELECT 1
-          FROM t_program_user u
-          WHERE u.c_program_id = $program_id
-          AND   u.c_user_id = $user_id
-          AND   u.c_role = 'pi'
-        )
-      """
-        .command
-        .contramap((caller, pid, uid) => (pid, uid, pid, caller))
-
-    val UnlinkObserverAsPiOrCoi: Command[(User.Id, Program.Id, User.Id)] =
-      sql"""
-        DELETE FROM t_program_user
-        WHERE c_program_id = $program_id
-        AND   c_user_id    = $user_id
-        AND   c_role       = 'coi_ro'
-        AND EXISTS (
-          SELECT 1
-          FROM t_program_user u
-          WHERE u.c_program_id = $program_id
-          AND   u.c_user_id    = $user_id
-          AND   (u.c_role = 'pi' OR u.c_role = 'coi')
-        )
-      """
-        .command
-        .contramap((caller, pid, uid) => (pid, uid, pid, caller))
-
-    /** Link a user to a program, without any access checking. */
-    val LinkUser: Fragment[(Program.Id, User.Id, UserType, ProgramUserRole, PartnerLink)] =
-      sql"""
-        INSERT INTO t_program_user (c_program_id, c_user_id, c_user_type, c_role, c_partner_link, c_partner)
-        SELECT $program_id, $user_id, $user_type, $program_user_role, $partner_link
-      """
+          c_program_user_id = $program_user_id
+      """.query(program_id *: program_user_role *: user_id.opt)
 
     /**
-     * Link a co-investigator to a program.
+     * Access checks for adding/updating a COI program user.
      * - Guests cannot do this.
      * - Staff, Admin, and Service users can always do this.
      * - Standard user can only do this if they're the program's PI.
```

Comment:
https://app.shortcut.com/lucuma/story/4178/cannot-invite-investigators-if-i-m-a-coi Considering this story, should COIs be able to add/update other COIs? Might be a question for Andy.